platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCollection.java

```diff
@@ -282,243 +336,244 @@ public boolean isLoadedFromExistingFiles() {
-                        + "of indexedFileList, so we should always have something here.";
-        final KeyRange keyRange = validKeyRange;
-        // open iterators, first iterator will be on oldest file
-        final List<DataFileIterator> blockIterators = new ArrayList<>(filesToMerge.size());
-        for (final DataFileReader<D> fileReader : filesToMerge) {
-            blockIterators.add(fileReader.createIterator());
-        }
-        // move all iterators to first block
-        ListIterator<DataFileIterator> blockIteratorsIterator = blockIterators.listIterator();
-        while (blockIteratorsIterator.hasNext()) {
-            final DataFileIterator dataFileIterator = blockIteratorsIterator.next();
-            try {
-                if (!dataFileIterator.next()) {
-                    // we have finished reading this file so don't need it iterate it next time
-                    dataFileIterator.close();
-                    blockIteratorsIterator.remove();
-                }
-            } catch (final IOException e) {
-                logger.error(EXCEPTION.getMarker(), "Failed while removing finished data file iterators", e);
-            }
+        // We need a map to find readers by file index below. It doesn't have to be synchronized
+        // as it will be accessed in this thread only, so it can be a simple HashMap or alike.
+        // However, standard Java maps can only work with Integer, not int (yet), so auto-boxing
+        // will put significant load on GC. Let's do something different
+        int minFileIndex = Integer.MAX_VALUE;
+        int maxFileIndex = 0;
+        for (final DataFileReader<D> r : filesToMerge) {
+            minFileIndex = Math.min(minFileIndex, r.getIndex());
+            maxFileIndex = Math.max(maxFileIndex, r.getIndex());
         }
-        // while we still have data left to read
-        long lastLowestKey = -1;
-        long[] thisRoundsKeys = new long[blockIterators.size()];
-        long[] lastRoundsKeys = new long[blockIterators.size()];
-        while (!blockIterators.isEmpty()) {
-            // find the lowest key any iterator has
-            long lowestKey = Long.MAX_VALUE;
-            for (int i = 0; i < blockIterators.size(); i++) {
-                final DataFileIterator blockIterator = blockIterators.get(i);
-                final long key = blockIterator.getDataItemsKey();
-                if (key < lowestKey) {
-                    lowestKey = key;
-                }
-                thisRoundsKeys[i] = key;
+        final int firstIndexInc = minFileIndex;
+        final int lastIndexExc = maxFileIndex + 1;
+        final DataFileReader<D>[] readers = new DataFileReader[lastIndexExc - firstIndexInc];
+        for (DataFileReader<D> r : filesToMerge) {
+            readers[r.getIndex() - firstIndexInc] = r;
+        }
+
+        final KeyRange keyRange = this.validKeyRange;
+        index.forEach((path, dataLocation) -> {
```

Comment:
This is a quite long lambda expression.  It might be easier to follow the design if the expression was extracted into a method or an inner class that implements BiFunction.

Reply:
The corresponding code before this fix (not lambda, but conceptually the same) is almost 100 lines, while the lambda is just 30 :) To me, these lines are very clear. If they are extracted to a separate method, the context would be lost and code will be less readable - although I'm quite biased here, I've been looking at this method for weeks.

Regarding BiFunction - this is intentionally done with a dedicated interface, as all standard interfaces in java.util can't work with primitive types. Previously Oleg observed very high number of memory allocations (very short, but still) from Long.valueOf() in this method, this is addressed now.

Reply:
BiFunction would replace the lambda expression with its concretized form, not change the actions within the lambda; there shouldn't be any additional allocations for that (and it shouldn't change the semantics, since lambda just creates an anonymous BiFunction anyway).
A method call wouldn't change the semantics much either, but at least it would be clearer to the reader where the lambda code starts and ends; which improves readability.

Not worth holding up this change, but something to think about as a readability improvement.